spec/core/ics-004-channel-and-packet-semantics/LOCALHOST.md

```diff
@@ -0,0 +1,72 @@
+
+Here is an example of how this would be implemented for the `recvPacket` handler:
+
+```typescript
+function recvPacket(packet: OpaquePacket,
+  proof: CommitmentProof,
+  proofHeight: Height,
+  relayer: string): Packet {
+    // pre-verification logic
+
+    // get the underlying connectionID for this channel
+    channel = provableStore.get(channelPath(packet.destPort, packet.destChannel))
+    connectionID = channel.connecti
+
+    //construct the expectedState and keypath
+    expectedState = commitPacket(packet)
+    keyPath = channelPath(packet.sourcePort, packet.sourceChannel, packet.Sequence)
+
+
+    verifyPacketCommitment(connectionID, keyPath, expectedState, proof, proofHeight)
+}
+
+// private function created to abstract the verification logic of localhost and remote chains from the rest of channel logic
+//
+// NOTE: It currently takes in a packet so we can call the connection verify function appropriately
+// however, a future refactor that abstracts verification at the connection layer
+// may remove the need for this additional argument.
+function verifyPacketCommitment(connectionID: string, packet: Packet, keyPath: bytes, expectedState: bytes, proof: CommitmentProof, proofHeight: Height) {
+    if connectionID == LocalHostConnection {
+        // verify packet by checking if the expected state 
+        // exists in our own state at the given keyPath
+        // NOTE: proof and proofHeight are ignored in this case
+        actualState = provableStore.get(keyPath)
+        abortTransactionUnless(expectedState == actualState)
+    } else {
+        abortTransactionUnless(connection !== null)
+        abortTransactionUnless(connection.state === OPEN)
+
+        abortTransactionUnless(connection.verifyPacketData(
+            proofHeight,
+            proof,
+            packet.sourcePort,
+            packet.sourceChannel,
+            packet.sequence,
+            expectedState,
+        ))
+    }
+}
+```
+
+## Relayer Messages
+
+Relayers supporting localhost packet flow must be adapted to submit messages from sending applications back to the originating chain.
+
+This would require first checking the underlying connectionID on any channel-level messages. If the underlying connectionID is `connection-localhost`, then the relayer must construct the message with an empty proof and proofHeight and submit the message back to the originating chain.
+
+Implementations **may** choose to implement localhost such that the next message in the handshake or packet flow is automatically called without relayer-driven transactions. However, implementors must take care to ensure that automatic message execution does not cause gas consumption issues.
```

Comment:
I'm curious what possible solutions exist for dealing with the gas consumption issue mentioned here. I know in the last ibc call the idea of using `EndBlock` for localhost packets was mentioned, would the same gas consumption solution for the relayer then be able to be applied to eliminate the need for relayers on localhost ibc connections?

Reply:
We could use EndBlock, but there are still block gas limits. Its possible it just would require some forethought (perhaps maintain a queue and limit the number of transactions that get processed per block). CCV already has some prior art on this i believe. leaning towards leaving this as-is in the spec, since its very implementation-specific

Reply:
I think some auto-call will be interesting, but a possible attack vector, and needs to be done very carefully.

But it definitely should be called in a different atomic transaction. My desire for such local host calls is to allow async passing and insulate from rollback/gas. For example, you could send "update messages" when something changed (ideally via pub sub) and listeners wouldn't add extra cost or DoS to the sender, as they need to pay own gas later (and making the issue of relayer incentivization even more pressing)

Anyway, just my two cents. I see you are going in this direction, and it is quite cool.